### PR TITLE
Improve error message when a plugin sets an invalid android package

### LINKED ISSUE #47803: _getSupportedEmbeddings claims kotlin file exists even when it doesn't
In `_getSupportedEmbeddings`, if neither the Java nor the Kotlin file exists, we report an error that says the Kotlin file exists but can't be read.

You can reproduce this by trying to use the `image_saver` plugin, because its Java file is in a different directory than the package name would suggest, and so the new code just fails to handle it entirely.

(See also #47802, which is a minor bugi n the same code.)

cc @blasten

### LINKED ISSUE #48165: Error in running any flutter commands related to app - couldn't read file ErrorHandlingFile even though it exists.
Error in trying to run Flutter packages get or flutter build 

**Target Platform:** Android
**Target OS version/browser:** Android 10
**Devices:** android studio android 10 emulator

## Logs

```
Running "flutter pub get" in appname...                              0.8s
Couldn't read file Instance of 'ErrorHandlingFile' even though it exists. Please verify that this file has read permission and try again.
Process finished with exit code 1
```

```
flutter doctor -v
[✓] Flutter (Channel master, v1.13.7-pre.24, on Mac OS X 10.15.2 19C57, locale
    en-GB)
    • Flutter version 1.13.7-pre.24 at
      /Users/harshapulikollu/Documents/development/flutter
    • Framework revision 0aed0b61a1 (5 hours ago), 2020-01-03 17:53:54 -0800
    • Engine revision eb139936eb
    • Dart version 2.8.0 (build 2.8.0-dev.0.0 2f57602411)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/harshapulikollu/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling
      support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3, Build version 11C29
    • CocoaPods version 1.8.4

[✗] Chrome - develop for the web (Cannot find chrome executable at
    /Applications/Google Chrome.app/Contents/MacOS/Google Chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 42.1.1
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build
      1.8.0_202-release-1483-b49-5587405)

[✓] VS Code (version 1.41.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.7.1

[✓] Connected device (1 available)
    • Web Server • web-server • web-javascript • Flutter Tools

! Doctor found issues in 1 category.
```

### PR DESCRIPTION
**Pull request by blasten**
## Description

If a plugin contains an invalid `androidPackage` entry in pubspec.yaml, then throw a tool exits with a useful/actionable error message. 

## Related Issues

Fixes https://github.com/flutter/flutter/issues/47803
Fixes https://github.com/flutter/flutter/issues/48165

## Tests

I added the following tests: Unit test.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_tools/lib/src/platform_plugins.dart [hunk 1]
```diff
@@ -84,33 +84,45 @@ class AndroidPlugin extends PluginPlatform {
...
+    bool mainClassFound = false;
+    for (String mainClassCandidate in mainClassCandidates) {
+      mainPluginClass = fs.file(mainClassCandidate);
+      if (mainPluginClass.existsSync()) {
+        mainClassFound = true;
+        break;
+      }
+    }
+    if (!mainClassFound) {
+      throwToolExit(
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/platform_plugins.dart]: assert `mainClassCandidates.length <= 2`

File: packages/flutter_tools/lib/src/platform_plugins.dart [hunk 2]
```diff
@@ -84,33 +84,45 @@ class AndroidPlugin extends PluginPlatform {
...
+        'The plugin $name doesn\'t have a main class defined in ${mainClassCandidates.join(' or ')}. '
+        'This is likely to due to an incorrect `androidPackage: $package` entry in the plugin\'s pubspec.yaml. '
+        'Please contact the author of this plugin and consider using a different plugin in the meanwhile. '
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/platform_plugins.dart]: The person getting the error is most likely to be the plugin author themselves. We should make sure to give useful actionable information to the plugin author as well.

File: packages/flutter_tools/test/general.shard/plugins_test.dart [hunk 3]
```diff
@@ -425,6 +455,32 @@ dependencies:
...
+        createPluginWithInvalidAndroidPackage();
+
+        await expectLater(
+          () async {
+            await injectPlugins(flutterProject);
+          },
+          throwsToolExit(
+            message: 'The plugin plugin1 doesn\'t have a main class defined in '
+                     '/.tmp_rand0/flutter_plugin_invalid_package.rand0/android/src/main/java/plugin1/invalid/UseNewEmbedding.java or '
+                     '/.tmp_rand0/flutter_plugin_invalid_package.rand0/android/src/main/kotlin/plugin1/invalid/UseNewEmbedding.kt. '
```
Comment (Hixie) [inline packages/flutter_tools/test/general.shard/plugins_test.dart]: this is probably going to fail on windows

File: packages/flutter_tools/lib/src/platform_plugins.dart [hunk 4]
```diff
@@ -84,33 +84,45 @@ class AndroidPlugin extends PluginPlatform {
...
     }
-    assert(mainPluginClass.existsSync());
+
     String mainClassContent;
     try {
       mainClassContent = mainPluginClass.readAsStringSync();
     } on FileSystemException {
       throwToolExit(
-        'Couldn\'t read file $mainPluginClass even though it exists. '
+        'Couldn\'t read file ${mainPluginClass.path} even though it exists. '
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/platform_plugins.dart]: looks like this isn't tested

**Comment by Hixie** [comment]
thanks!

File: packages/flutter_tools/lib/src/platform_plugins.dart (hunk 4 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/platform_plugins.dart]: done :)

File: packages/flutter_tools/lib/src/platform_plugins.dart (hunk 1 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/platform_plugins.dart]: 👍done

File: packages/flutter_tools/lib/src/platform_plugins.dart (hunk 2 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/platform_plugins.dart]: sounds good. I added this to the message: "If you are the author of this plugin, fix the `androidPackage` entry or move the main class to any of locations used above."

File: packages/flutter_tools/test/general.shard/plugins_test.dart (hunk 3 above)
Comment (blasten) [reply packages/flutter_tools/test/general.shard/plugins_test.dart]: It looks like the `MemoryFileSystem` defaults to `/` :)

**Review by kakaka11 — APPROVED** [review]

**Comment by blasten** [comment]
PTAL @Hixie

**Review by zanderso — APPROVED** [review]

**Pull request merged**